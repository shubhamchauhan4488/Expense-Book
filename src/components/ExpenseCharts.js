// import CanvasJS, {pie} from 'canvasjs';
import React, {Component} from "react";
import {Bar, Line, Pie} from "react-chartjs-2";
import getVisibleExpenses from '../selectors/expenses'
import { connect } from "react-redux";


class ExpenseCharts extends Component {	

	constructor(props){
		super(props)

		this.state = {
			householdSum : '',
			utilitiesBillsSum : '',
			rentSum : '',
			entertainmentSum : '',
			healthSum : '',
			transportSum : '',
			educationSum : '',
			miscellaneousSum : ''
		}
		this.setChartData = this.setChartData.bind(this)
	}

	setChartData(props){

		let householdSum = props.expenses.filter((expense) => {
			return expense.category === 'Household'
		}).map((expense) =>  expense.amount)
		.reduce((sum, value) =>  sum + value, 0)

		let utilitiesBillsSum = props.expenses.filter((expense) => {
			return expense.category === 'Utilities Bill'
		}).map((expense) =>  expense.amount)
		.reduce((sum, value) =>  sum + value, 0)

		let rentSum = props.expenses.filter((expense) => {
			return expense.category === 'Rent'
		}).map((expense) =>  expense.amount)
		.reduce((sum, value) =>  sum + value, 0)
		
		let entertainmentSum = props.expenses.filter((expense) => {
			return expense.category === 'Entertainment'
		}).map((expense) =>  expense.amount)
		.reduce((sum, value) =>  sum + value, 0)
	
		let healthSum = props.expenses.filter((expense) => {
			return expense.category === 'Health'
		}).map((expense) =>  expense.amount)
		.reduce((sum, value) =>  sum + value, 0)

		let transportSum = props.expenses.filter((expense) => {
			return expense.category === 'Transport'
		}).map((expense) =>  expense.amount)
		.reduce((sum, value) =>  sum + value, 0)

		let educationSum = props.expenses.filter((expense) => {
			return expense.category === 'Education'
		}).map((expense) =>  expense.amount)
		.reduce((sum, value) =>  sum + value, 0)

		let miscellaneousSum = props.expenses.filter((expense) => {
			return expense.category === 'Miscellaneous'
		}).map((expense) =>  expense.amount)
		.reduce((sum, value) =>  sum + value, 0)

		this.setState({
			householdSum : householdSum/100,
			utilitiesBillsSum : utilitiesBillsSum/100,
			rentSum : rentSum/100,
			entertainmentSum : entertainmentSum/100,
			healthSum : healthSum/100,
			transportSum : transportSum/100,
			educationSum : educationSum/100,
			miscellaneousSum : miscellaneousSum/100
		})
	}

	componentWillReceiveProps(nextProps) {
		this.setChartData(nextProps)
	}

	componentDidMount(){
		this.setChartData(this.props)
	}

	render() {

		const chartData = {
						labels : ['Household', 'Utilities Bill', 'Rent', 'Entertainment', 'Health', 'Transport','Education', 'Miscellaneous'],
						datasets  : [ {
				
							data : [
								this.state.householdSum, 
								this.state.utilitiesBillsSum, 
								this.state.rentSum,
								this.state.entertainmentSum,
								this.state.healthSum,
								this.state.transportSum,
								this.state.educationSum,
								this.state.miscellaneousSum],
							backgroundColor : [
								'rgba(221,28,28,0.6)',
								'rgba(77,218,234,0.6)',
								'rgba(52, 122, 42,0.6)',
								'rgba(242, 238, 116,0.6)',
								'rgba(200, 116, 242,0.6)',
								'rgba(219, 168, 189,0.6)',
								'rgba(255,45,152,0.6)',
								'rgba(255,9,132,0.6)'
							]
						 }
						]
					};
		return (
			<div className = "charts">
				<Bar 
				data = {chartData}
				options = {{
					title : {
						display : true,
						text : "Category wise expenses",
						fontSize : 25,
						margin: 20
					},
					legend : {
						display : false,
						position: 'left',
						labels: {
						boxWidth: 10
						}
					}
				}}
				/>
			</div>
		)
	}
}
		   
const mapStateToProps = (state) => {
	return{
		expenses : getVisibleExpenses(state.expenses, state.filters)
	}
}
	
 export default connect(mapStateToProps)(ExpenseCharts);