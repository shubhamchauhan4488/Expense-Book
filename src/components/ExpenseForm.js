import React from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";

const now = moment();
console.log(now.format("MMMM Do, YYYY"));

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    //While testing moment() will result in a different result from what was in the previos snapshot,
    //So testing willl fail , hence we mock moment.js
    this.state = {
      description: props.expense ? props.expense.description : "",
      note: props.expense ? props.expense.note : "",
      amount: props.expense ? (props.expense.amount / 100).toString() : "",
      category: props.expense ? props.expense.category : "Household",
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calenderFocused: false,
      error: ""
    };
    this.isSelected = this.isSelected.bind(this);
  }

  onDescriptionChange = e => {
    const description = e.target.value;
    this.setState(() => {
      return {
        description: description
      };
    });
  };
  onNoteChange = e => {
    // e.persist();
    // this.setState(() => {
    //     return{
    //         note : e.target.value
    //     }
    // })

    // OR
    const note = e.target.value;
    this.setState(() => {
      return {
        note: note
      };
    });
  };
  onAmountChange = e => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      //d{1,} will prevent us from entering numbers 0 to 1 i.e 0.122
      this.setState(() => {
        return {
          amount: amount
        };
      });
    }
  };
  onCategoryChange = e => {
    const category = e.target.value;
    this.setState(() => {
      return {
        category: category
      };
    });
  };
  onDateChange = createdAt => {
    this.setState(() => {
      if (createdAt) {
        return {
          createdAt: createdAt
        };
      }
    });
  };
  onFocusChange = ({ focused }) => {
    this.setState(() => {
      return {
        calenderFocused: focused
      };
    });
  };
  onSubmit = e => {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      this.setState(() => {
        return {
          error: "Please enter both description and amount"
        };
      });
    } else {
      this.setState(() => {
        return {
          error: ""
        };
      });
      //accessing the onSubmit function which is passed as props from <addexpense> and
      //passing the new expense to the <Addexpense> so that it can be added to redux store
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        note: this.state.note,
        category: this.state.category,
        createdAt: this.state.createdAt.valueOf()
      });
    }
  };
  isSelected(category) {
    if (this.state.category) {
      return this.state.category === category ? true : false;
    }
  }

  render() {
    return (
      <form className="form" onSubmit={this.onSubmit}>
        {this.state.error && <p className="form-error">{this.state.error}</p>}

        <input
          type="text"
          className="text-input"
          placeholder="Description"
          autoFocus
          value={this.state.description}
          onChange={this.onDescriptionChange}
        />

        <input
          type="text"
          className="text-input"
          placeholder="Amount"
          value={this.state.amount}
          onChange={this.onAmountChange}
        />

        <select className="text-input" onChange={this.onCategoryChange}>
          <option value="Household" selected={this.isSelected("Household")}>
            Household
          </option>
          <option value="Rent" selected={this.isSelected("Rent")}>
            Rent
          </option>
          <option
            value="Utilities Bill"
            selected={this.isSelected("Utilities Bill")}
          >
            Utility Bills
          </option>
          <option
            value="Entertainment"
            selected={this.isSelected("Entertainment")}
          >
            Entertainment
          </option>
          <option value="Health" selected={this.isSelected("Health")}>
            Health
          </option>
          <option value="Transport" selected={this.isSelected("Transport")}>
            Transport
          </option>
          <option value="Education" selected={this.isSelected("Education")}>
            Education
          </option>
          <option
            value="Miscellaneous"
            selected={this.isSelected("Miscellaneous")}
          >
            Miscellaneous
          </option>
        </select>

        <SingleDatePicker
          date={this.state.createdAt}
          onDateChange={this.onDateChange}
          focused={this.state.calenderFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
        <textarea
          className="textarea-input"
          placeholder="Add a note for this expense (optional)"
          value={this.state.note}
          onChange={this.onNoteChange}
        />
        <div>
          <button className="button">Save Expense</button>
        </div>
      </form>
    ); //button put inside div to exclude it from the styling of the direct descendants of form
  }
  // isOutsideRange = {() => false} takes day as an argument, but we want to show every day so we just return false
  //numberOfMonths = {1} means how many months u want to show on the datepicker
}
