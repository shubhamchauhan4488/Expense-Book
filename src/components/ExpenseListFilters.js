import React from 'react'
import {connect} from 'react-redux'
// import filters from '../reducers/filters';
import { setTextFilter, sortByDate, sortByAmount } from '../actions/filters';


const ExpenseListFilters = (props) => (
    <div>
        <input 
        type = "text" 
        value = {props.filters.text} 
        onChange = {(e) => {
            props.dispatch(setTextFilter(e.target.value))
        }}
        />  
        <select
        value = {props.filters.sortBy}
        onChange = {(e) => {
        if(e.target.value === 'date'){
            props.dispatch(sortByDate())
        }
        else if(e.target.value === 'amount'){
            props.dispatch(sortByAmount())
        }
        }}>
            <option value = "date" >Date</option>   
            <option value = "amount">Amount</option>  
        </select>
    </div>
    // option value = "date" means we are passing this value 'date' instead of 'Date'
)
const mapStateToProps = (state) => {
    console.log(state.filters)
    return {
        filters : state.filters
    }
}

const ConnectedExpenseListFilters = connect(mapStateToProps)(ExpenseListFilters)
export default ConnectedExpenseListFilters