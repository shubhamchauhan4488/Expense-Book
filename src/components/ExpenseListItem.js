import React from 'react'
import { Link } from "react-router-dom";
import moment from 'moment'
import numeral from 'numeral'

export const ExpenseListItem = ({ dispatch, id, description, amount, createdAt}) => (
    <Link className = "list-item" to = {`/edit/${id}`}>
    <div>
        <h3 className = "list-item__title">{description}</h3>
        <span className = "list-item__subtitle">{moment(createdAt).format('MMMM Do,YYYY')}</span>
    </div>
        <h3>{numeral(amount/100).format('$0,0.00')}</h3>
    </Link>
)

export default ExpenseListItem
//we are exporting the connected component 
//but here we dont need access to the state so we can use only connect(), it will give us access to dispatch prop