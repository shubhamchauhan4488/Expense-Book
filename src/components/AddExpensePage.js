import React from 'react';
import ExpenseForm from './ExpenseForm'
import { connect } from "react-redux";
import { addExpense } from '../actions/expenses';


// const AddExpensePage = (props) => (
//     <div>
//     This is from AddExpensePage
//     <ExpenseForm 
//     onSubmit = {(expense) => {
//         console.log(expense)
//         props.dispatch(addExpense(expense));
//         props.history.push('/')
//     }}
//     />
//     </div>
// )
// // props.history.push('/') : to switch to pages without reloading the page
// export default connect()(AddExpensePage)

//For testing we will modify the component : New class component:
export class AddExpensePage extends React.Component{
    onSubmit = (expense) => {
        console.log(expense)
        this.props.addExpense(expense) //calling the addExpense inside mapDispatchToProps for dispatch handling
        this.props.history.push('/')
    }
    render(){
        return  (
            <div>
            This is from AddExpensePage
            <ExpenseForm 
            onSubmit = {this.onSubmit}
            />
            </div>
        )
    }
}

//this function wwe will use to handle the dispatch here itself, instead of handling at:
// props.dispatch(addExpense(expense));
// we do this b'cause while testing we will have to pass spy fn
// and in this case we have to pass a component in a spy fn which will come out ot be very complex
const mapDispatchToProps = (dispatch) => ({
    addExpense : (expense) => dispatch(addExpense(expense))
})

export default connect(undefined, mapDispatchToProps)(AddExpensePage)
