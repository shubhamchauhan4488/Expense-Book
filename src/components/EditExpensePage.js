import React from "react";
import ExpenseForm from "./ExpenseForm";
import { startEditExpense } from "../actions/expenses";
import { connect } from "react-redux";
import { startRemoveExpense } from "../actions/expenses";

export class EditExpensePage extends React.Component {
  onSubmit = expense => {
    this.props.startEditExpense(this.props.expense.id, expense);
    this.props.history.push("/");
  };
  onRemove = () => {
    this.props.startRemoveExpense(this.props.expense.id);
    this.props.history.push("/");
  };
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="container">
            <h1 className="page-header__title">Edit Expense</h1>
          </div>
        </div>
        <div className="container">
          <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit} />
          <button className="button button-secondary" onClick={this.onRemove}>
            Remove Expense
          </button>
        </div>
      </div>
    );
  }
}

//below props are the ones which are passed to the HOC,
//nd we are already aware that we can pass more props to the wrapped compoent : so that is what we will do
const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find(expense => {
      return props.match.params.id === expense.id;
    })
  };
};
const mapDispatchToProps = (dispatch, props) => ({
  startRemoveExpense: data => {
    dispatch(startRemoveExpense(data));
  },
  startEditExpense: (id, expense) => {
    dispatch(startEditExpense(id, expense));
  }
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
