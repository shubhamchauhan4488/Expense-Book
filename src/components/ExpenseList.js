import React from 'react'
import { connect } from 'react-redux'
import ExpenseListItem from './ExpenseListItem'
import getVisibleExpenses from "../selectors/expenses";

//we are exporting this separately b'cause we need an un-connected component for testing, it should not be connected to the store
export const ExpenseList = (props) => (
    <div className = "container">
        <div className = "list-header">
            <div className = "show-for-mobile">Expenses</div>
            <div className = "show-for-desktop">Expense</div>
            <div className = "show-for-desktop">Amount</div>
        </div>
        <div className = "list-content">
            {props.expenses.length === 0 ? (
                <div className = "list-item list-item--error">
                  <p>No Expenses to display</p>
                </div>
            ) : (
                props.expenses.map((expense) => {
                    return(
                        <ExpenseListItem key = {expense.id} {...expense}></ExpenseListItem>
                )
                })
            )}
        </div>
    </div>
)

//Since this is connected to the store, we will always get the latest values of the state
const mapStateToProps = (state) => {
    // console.log(state.filters)
        return {
            expenses : getVisibleExpenses(state.expenses, state.filters)
        }
    }
const ConnectedExpenseList = connect(mapStateToProps)(ExpenseList)

export default ConnectedExpenseList;

// const ConnectedExpenseList = connect((state) => {
//     return {
//         expenses : state.expenses,
//         filters : state.filters
//     }
// })(ExpenseList)
// export default ConnectedExpenseList;

//connect() api wants us to specify wrappedComponent as connect()(WrappedComponent) here connect()(ExpenseList)

//Passing props 
//Example to the HOC like :  // ReactDOM.render(<ReturnedAuthHOC requireAuth = {true} info = "These are the details in this document"/>, document.getElementById('main-div'))
//Here we have to pass those to connect() and ConnectedExpenseList is the HOC, Now we will pass it using a callback function that returns 
//object (in key value pair) and we can select selective thing from the store 

//Now this key value pair can be accessed by the ExpenseList which was wrapped 
//const ExpenseList = (props) => (
//     <div>
//         <h3>
//             {props.expenes}
//         </h3>
//     </div>
// )



// const requireAuthentication = (WrappedComponent) => {
//     return (props) => (
//         <div>
//             {props.requireAuth ? (
//                 <WrappedComponent {...props}/>
//             ):
//             <p>Please login correctly to view the Info</p>}
//         </div>
//     )
// }
// const ReturnedAuthHOC = requireAuthentication(Info)
