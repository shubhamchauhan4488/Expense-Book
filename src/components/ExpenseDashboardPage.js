import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpensesSummary from './ExpensesSummary';
import ExpenseCharts from './ExpenseCharts'

const ExpenseDashboardPage = () => (
    <div>
        <ExpensesSummary />
        <ExpenseListFilters />
        <ExpenseList/>
        <ExpenseCharts/>
    </div>
)

export default ExpenseDashboardPage;