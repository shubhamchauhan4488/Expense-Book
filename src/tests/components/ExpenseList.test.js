import React from 'react'
import {ExpenseList}  from "../../components/ExpenseList";
import {shallow} from 'enzyme'
import expensesArray from "../testData/expenses";

test("should correctly render Expense LIst with given expenses", () => {
    const wrapper = shallow(<ExpenseList expenses = {expensesArray}/>);
    expect(wrapper).toMatchSnapshot()
})

test("should correctly render Expense LIst with NO expenses PASSED", () => {
    const wrapper = shallow(<ExpenseList expenses = {[]}/>);
    expect(wrapper).toMatchSnapshot()
})

