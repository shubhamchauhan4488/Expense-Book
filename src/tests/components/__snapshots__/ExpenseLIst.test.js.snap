// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render Expense LIst with NO expenses PASSED 1`] = `
<div
  className="container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Expenses
    </div>
    <div
      className="show-for-desktop"
    >
      Expense
    </div>
    <div
      className="show-for-desktop"
    >
      Amount
    </div>
  </div>
  <div
    className="list-content"
  >
    <div
      className="list-item list-item--error"
    >
      <p>
        No Expenses to display
      </p>
    </div>
  </div>
</div>
`;

exports[`should correctly render Expense LIst with given expenses 1`] = `
<div
  className="container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Expenses
    </div>
    <div
      className="show-for-desktop"
    >
      Expense
    </div>
    <div
      className="show-for-desktop"
    >
      Amount
    </div>
  </div>
  <div
    className="list-content"
  >
    <ExpenseListItem
      amount={195}
      createdAt={0}
      description="Gum"
      id="1"
      key="1"
      note="asd"
    />
    <ExpenseListItem
      amount={131245}
      createdAt={-345600000}
      description="Rent"
      id="2"
      key="2"
      note=""
    />
    <ExpenseListItem
      amount={4500}
      createdAt={345600000}
      description="Credit Card"
      id="3"
      key="3"
      note=""
    />
  </div>
</div>
`;
